temporalio/worker/_workflow.py

```diff
@@ -250,9 +316,32 @@ async def _handle_activation(
                         activate_task, self._deadlock_timeout_seconds
                     )
                 except asyncio.TimeoutError:
-                    raise RuntimeError(
-                        f"[TMPRL1101] Potential deadlock detected, workflow didn't yield within {self._deadlock_timeout_seconds} second(s)"
-                    )
+                    # extract the stack of the workflow if possible
+
+                    # cases:
+                    # 1. get_thread_id() is not implemented for the worker (returns None, fall to case 2)
+                    # 2. get_thread_id() returns None
+                    # 3. tid is not present in sys._current_frames()
+                    # 4. error occurs during traceback construction
+                    # 5. normal
+
+                    msg = f"[TMPRL1101] Potential deadlock detected: workflow didn't yield within {self._deadlock_timeout_seconds} second(s)."
+                    tid = workflow.get_thread_id()
+                    if not tid:
+                        # Default error if the function is not implemented for the workers, or the worker is not running
+                        raise RuntimeError(msg) from None
+
+                    tb, warn = get_traceback_for_thread(tid)
+
+                    if not tb:
+                        msg += " Worker was unable to retrieve workflow thread info. No frames from workflow available."
+                        raise RuntimeError(msg) from None
+
+                    if warn:
+                        msg += " System encountered errors while constructing traceback, minimal or partial traceback will be presented."
+
+                    raise RuntimeError(msg).with_traceback(tb) from None
```

Comment:
@twin-drill and I were also discussing whether we really want to present the captured stack trace as an error traceback when it doesn't actually correspond to an exception being raised. I'm thinking that the correct thing to do is present it as an informational message that contains a stack trace.

Reply:
It is a stacktrace in other langs and it kinda represents where the deadlock currently is and we have mechanisms to display stack traces in UI and others vs really long error messages. I personally like it here (the actual stack trace inside our code is not meaningful IMO)